src/main/java/frc/robot/subsystems/Grabber.java

```diff
@@ -0,0 +1,123 @@
+      case STOPPED: endEffectorSubsystem.stopMotors(); break;
+    }
+    efMode = effectorMode;
+  }
+
+  public double[] getArmPositions(ArmModes armMode){
+    switch (armMode){
+      case INTAKING_CUBE: return ArmConstants.intakingCubesArmPos;
+      case INTAKING_CONE_UPRIGHT: return ArmConstants.intakingConesUprightArmPos;
+      case INTAKING_CONE_FALLEN: return ArmConstants.intakingConesFallenArmPos;
+      case HOLDING: return ArmConstants.holdingArmPos;
+
+      case PLACING_CONE_LVL1: return ArmConstants.placingConeArmPosOne;
+      case PLACING_CONE_LVL2: return ArmConstants.placingConeArmPosTwo;
+      case PLACING_CONE_LVL3: return ArmConstants.placingConeArmPosThree;
+
+      case PLACING_CUBE_LVL1: return ArmConstants.placingCubeArmPosOne;
+      case PLACING_CUBE_LVL2: return ArmConstants.placingCubeArmPosTwo;
+      case PLACING_CUBE_LVL3: return ArmConstants.placingCubeArmPosThree;
+    }
+    return new double[]{};
+  }
+
+  private EFModes efMode = EFModes.STOPPED;
+  private ArmModes armMode = ArmModes.HOLDING;
+  private EFModes desiredEFMode = EFModes.STOPPED;
+  private ArmModes desiredArmMode = ArmModes.HOLDING;
+
+  public void setDesiredEFMode(EFModes desiredEFMode){
+    this.desiredEFMode = desiredEFMode;
+  }
+  public void setDesiredArmMode(ArmModes desiredArmMode){
+    this.desiredArmMode = desiredArmMode;
+  }
+  public EFModes getDesiredEFMode(){
+    return desiredEFMode;
+  }
+  public ArmModes getDesiredArmMode(){
+    return desiredArmMode;
+  }
+
+  public EFModes getEFMode(){
+    return efMode;
+  }
+  public ArmModes getArmMode(){return armMode;}
+
+
+
+
+  public void setArmPosition(double[] armPosition){
+    if(armPosition.length<2) return;
+    armStageOne.setArmPositionRad(armPosition[0]);
+    armStageOne.setArmPositionRad(armPosition[1]);
+  }
+
+  public void setDesiredArmAndEFModes(ArmModes armMode, EFModes efMode){
+    setDesiredEFMode(efMode);
+    setDesiredArmMode(armMode);
+  }
+
+
+  @Override
+  public void periodic() {
+    // This method will be called once per scheduler run
+
+    if(getEFMode() != getDesiredEFMode()){ //wait for the arm to be within the tolerance to update the EF mode if intaking or placing
+      boolean armOneAligned = MathThings.isWithinRangeOf(armStageOne.getEncoderRad(), getArmPositions(getDesiredArmMode())[0],ArmConstants.angleToleranceToUpdateEF);
+      boolean armTwoAligned = MathThings.isWithinRangeOf(armStageTwo.getEncoderRad(), getArmPositions(getDesiredArmMode())[1],ArmConstants.angleToleranceToUpdateEF);
+      boolean desiredEFIsHolding = getDesiredEFMode()==EFModes.HOLDING_CONE || getDesiredEFMode()==EFModes.HOLDING_CUBE;  //don't wait if you're switching to a holding mode
+      if((armOneAligned && armTwoAligned) || (desiredEFIsHolding)){
+        setEndEffectorMode(desiredEFMode);
+      }
+    }
+
+    if(getArmMode()!=getDesiredArmMode()){ //instantly set arm position if it doesn't match the desired position
```

Comment:
do we need a desired arm mode if our arm mode and desired are always the same? are they not? is it just for congruency between arm and EF, cus if so that's acceptable.

Reply:
Yeah, it's mostly just for congruency